src/transactions.rs

```diff
@@ -16,17 +16,20 @@ use crate::{
 };
 use atomic_store::{AppendLog, AtomicStoreLoader};
 use chrono::{DateTime, Local};
+use derivative::Derivative;
 use jf_cap::{
     keys::UserAddress,
     structs::{ReceiverMemo, RecordOpening},
     Signature,
 };
-use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use serde::{Deserialize, Serialize};
 use std::collections::HashSet;
 use std::ops::{Deref, DerefMut};
 
 /// A Transaction<L>with its UID as the primary key.
-#[derive(Clone, Debug, Deserialize, Serialize)]
+#[derive(Clone, Debug, Derivative, Deserialize, Serialize)]
+#[serde(bound = "")]
+#[derivative(PartialEq(bound = "L: Ledger"))]
 pub struct Transaction<L: Ledger> {
```

Comment:
Looking over the rest of this code, I find myself getting a little confused occasionally about which fields are set for received transactions, sent transactions, observed transactions....

Do you have any ideas about how we could clarify this? A good starting point would just be to include in the Rust doc for each field the conditions under which it is present. We might be able to do better, at least for some fields, by having sub-structures like `SentTransaction` for the sender-only fields and `ReceivedTransaction`, etc. Then in `Transaction`, instead of a bunch of related-by-convention fields, we could just have like `sender_info: Option<SentTransaction>`

Reply:
The only way I can think of is to have optional send and receive structs and optional fields for them in `Transactions`.

Reply:
Yeah. Should we table this for future work?